### PR TITLE
feat: fix dos memory

### PR DESCRIPTION
**Pull request by anudeeps0306**
---

## Description

This PR fixes critical memory exhaustion vulnerabilities in the backend that could cause server crashes or Denial of Service when handling large CSV datasets.

### Key Improvements

1. **Refactored `preprocess_data` (app/services/data_process.py)**

   * Replaced `pandas.read_csv()` (full in-memory loading) with `dask.dataframe.read_csv()`.
   * Implemented out-of-core processing using Dask partitions.
   * Applied transformations (Min-Max Normalization, One-Hot Encoding, Log Transform) using lazy evaluation.
   * Ensures large datasets are processed incrementally without exhausting server memory.

2. **Refactored `model_run` (app/services/model_run.py)**

   * Replaced full in-memory pandas arrays for training with `tf.data.experimental.make_csv_dataset`.
   * Enabled streaming CSV batches directly from disk to TensorFlow.
   * Eliminated memory spikes during deep learning model training.

3. **Test Suite Updates**

   * Updated tests to mock `dask.dataframe.read_csv` and Dask `.to_csv()` methods instead of pandas.
   * Fixed Dask One-Hot Encoding issue by explicitly categorizing columns.
   * Added missing Dask dependency configuration for test environment.

All changes are committed to the `fix-dos-memory` branch.

Fixes #<issue_number>

---

## Type of Change

* [x] Bug fix
* [ ] New feature
* [ ] Breaking change
* [ ] Documentation update

---

## How Has This Been Tested?

* Ran full backend Pytest suite (109 tests).

* Verified file uploads, transformations, model training flows, and API endpoints.

* Updated and validated mocks for Dask and TensorFlow streaming pipelines.

* Confirmed no memory spikes during large dataset simulation.

* [x] Existing tests pass

* [ ] New tests added

* [x] Manual testing

---

## Screenshots (if applicable)

Screenshot_2026-03-01_18-31-26

Screenshot_2026-03-01_18-31-54

## Checklist

* [x] My code follows the project's style guidelines
* [x] I have performed a self-review
* [x] I have added/updated documentation as needed
* [x] My changes generate no new warnings
* [x] Tests pass locally

---

### CODE REVIEW

File: tensormap-backend/app/services/model_run.py [hunk 1]
```diff
@@ -157,21 +156,42 @@ def _run(model_name: str, db: Session) -> None:
...
+            shuffle=True,
+            shuffle_seed=42,
+        )
+
+        # Pack the feature dict into a single feature array
+        def pack_features_vector(features, labels):
+            features = tf.stack(list(features.values()), axis=1)
+            return features, labels
+
+        dataset = dataset.map(pack_features_vector)
```
[the remark this answers is not in the document]
Comment (anudeeps0306) [reply tensormap-backend/app/services/model_run.py]: fixed them

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR makes a genuine and well-intentioned effort to fix memory-exhaustion (DoS) vulnerabilities by migrating from full in-memory pandas loading to Dask (lazy, out-of-core) and TensorFlow's streaming CSV reader. However, it introduces a **critical self-defeating bug** (`dataset.cache()` reloads the entire dataset into memory), several high-severity logic errors (incorrect `total_rows` calculation, silent `Log Transform` behavioral change, full Dask scan for row count), and a number of medium-severity code quality and correctness concerns. The test suite updates reflect the right approach but have fragile mock targets. The PR needs revisions before merging, particularly addressing the cache bug and the row-count correctness issues.

File: tensormap-backend/app/services/data_process.py [hunk 2]
```diff
@@ -230,7 +244,15 @@ def get_file_data(db: Session, file_id: uuid_pkg.UUID) -> tuple:
     return _resp(200, True, "Data sent successfully", data_json)
 
 
-_VALID_TRANSFORMATIONS = {"One Hot Encoding", "Categorical to Numerical", "Drop Column"}
+_VALID_TRANSFORMATIONS = {
+    "One Hot Encoding",
+    "Categorical to Numerical",
+    "Drop Column",
+    "Min-Max Normalization",
+    "Z-score Standardization",
```
Comment (ivantha) [inline tensormap-backend/app/services/data_process.py]: **[INFO]** Positive: `_VALID_TRANSFORMATIONS` correctly expanded to include all supported transformation names

The PR expands the validation set to include the four new transformations:
```python
_VALID_TRANSFORMATIONS = {
    "One Hot Encoding",
    "Categorical to Numerical",
    "Drop Column",
    "Min-Max Normalization",
    "Z-score Standardization",
    "Log Transform",
    "Fill Missing Values",
}
```

This ensures the validation gate at the start of `preprocess_data` correctly rejects unknown transformations and lists all valid options in the error message. Without this, the new transformations would silently pass validation but never execute (falling through all `if` checks).

**Suggestion:**
```
No change needed. Consider replacing the sequential `if` statements in the loop with an early `continue` guard or a dispatch dict to make the handler more maintainable as the number of transformations grows.
```

**Comment by ivantha** [comment]
This branch has merge conflicts with `main`. Please rebase onto the latest `main`, resolve the conflicts, and push an update. Also note the PR body still has the literal placeholder `Fixes #<issue_number>` — please update it to reference the correct issue (or remove the Fixes line if none applies).

**Comment by ivantha** [comment]
Thank you for identifying the DoS memory exhaustion vulnerability! This PR has accumulated change requests and conflicts with current main. The security improvements in this area have been partially addressed in other merged work. We encourage you to open a fresh, rebased PR focusing specifically on the streaming/pagination fix for large CSV loads — that contribution would be very welcome.
